src/python/pants/reporting/json_reporter.py

```diff
@@ -0,0 +1,141 @@
+
+  def open(self):
+    """Implementation of Reporter callback."""
+
+    safe_mkdir(os.path.dirname(self.report_path))
+
+  def close(self):
+    """Implementation of Reporter callback."""
+
+    mode = 'w' if PY3 else 'wb'
+
+    safe_file_dump(
+      self.report_path,
+      json.dumps({
+        'workunits': self._results,
+        'artifact_cache_stats': self.run_tracker.artifact_cache_stats.get_all(),
+      }),
+      mode=mode)
+
+  def start_workunit(self, workunit):
+    """Implementation of Reporter callback."""
+
+    workunit_data = {
+      'name': workunit.name,
+      'id': str(workunit.id),
+      'parent_name': workunit.parent.name if workunit.parent else '',
+      'parent_id': str(workunit.parent.id) if workunit.parent else '',
+      'labels': list(workunit.labels),
+      'cmd': workunit.cmd or '',
+      'start_time': workunit.start_time,
+      'outputs': defaultdict(str),
+      'children': [],
+      'log_entries': [],
+    }
+
+    root_id = str(workunit.root().id)
+
+    if not root_id in self._root_id_to_workunit_stack:
+      self._root_id_to_workunit_stack[root_id].append(workunit_data)
+      self._results[root_id] = workunit_data
+    else:
+      self._root_id_to_workunit_stack[root_id][-1]['children'].append(workunit_data)
+      self._root_id_to_workunit_stack[root_id].append(workunit_data)
+
+  def end_workunit(self, workunit):
+    """Implementation of Reporter callback."""
+
+    additional_data = {
+      'outcome': workunit.outcome_string(workunit.outcome()),
+      'end_time': workunit.end_time,
+      'unaccounted_time': workunit.unaccounted_time(),
+    }
+
+    root_id = str(workunit.root().id)
+
+    # We update the last workunit in the stack, which is always the
+    # youngest workunit currently open, and then pop it off. The
+    # last workunit serves as a pointer to the child workunit under
+    # it's parent (the parent of the youngest workunit currently
+    # open is always the one immediately above it in the stack).
+    self._root_id_to_workunit_stack[root_id][-1].update(additional_data)
+    self._root_id_to_workunit_stack[root_id].pop()
+
+  def handle_output(self, workunit, label, stream):
+    """Implementation of Reporter callback."""
+
+    self._root_id_to_workunit_stack[str(workunit.root().id)][-1]['outputs'][label] += stream
+
+  def do_handle_log(self, workunit, level, *msg_elements):
+    """Implementation of Reporter callback."""
+
+    entry_info = {
+      'level': self._log_level_str[level],
+      'messages': self._render_messages(*msg_elements),
+    }
+
+    root_id = str(workunit.root().id)
+    current_stack = self._root_id_to_workunit_stack[root_id]
+    if current_stack:
+      current_stack[-1]['log_entries'].append(entry_info)
+    else:
+      self._results[root_id]['log_entries'].append(entry_info)
+
+  def _render_messages(self, *msg_elements):
+    def _message_details(element):
+      if isinstance(element, string_types):
+        element = [element]
```

Comment:
This part is a little confusing to me, specifically because I don't know what the type of `*msg_elements` might be? Is it like this?
```python
JsonPrimitive = Union[str, bytes, int, float]
JsonElement = Union[JsonPrimitive, Dict[str, JsonElement], List[JsonElement]]
```

We can't define this in type hints yet, but think there are other ways to express this. For example, add some docstring, or add more `ifinstance` checks where relevant, such as 
```python
if ininstance(element, str):
  element = [element]
elif not isinstance(element, (int, list, dict):
  raise ValueError("Invalid element.")
```

Reply:
Got it, this was another part that came over from the `HtmlReporter` - and it seems how we handle message elements in `do_handle_log` in similar (in that they use `six.string_types`) across all reporters so far.

Per `report.py`, it looks like this should either be a `tuple` or a `str`:
https://github.com/pantsbuild/pants/blob/136f9cf86dd0fa4ce1fd979f553c67553331a953/src/python/pants/reporting/report.py#L100-L107

In which case it seems like it would be simpler to invert that logic and make `element` a `tuple` if it's _not_ already and drop `string_types` entirely? Something like:

```python
if not isinstance(element, tuple):
  element = (element,)
```

Not sure if that might risk running into other issues though... thoughts?

Reply:
Hm, if this is how we handle reporting in every other reporter, then I won't block and it would be outside the scope to fix here.

Instead, I added a task to https://github.com/pantsbuild/pants/issues/6071 for us to clean this up.

You can keep everything as is!